File: Raster.Ts.Dependencies/Modules/system.js

var _cpp = cpp;

exports.cpuCacheLineSize = function() {
    return _cpp.cpuCacheLineSize();
}

exports.cpuCount = function() {
    return _cpp.cpuCount();
}

exports.systemRam = function() {
    return _cpp.systemRam();
}

exports.has3DNow = function() {
    return _cpp.has3DNow();
}

exports.hasAVX = function() {
    return _cpp.hasAVX();
}

exports.hasAVX2 = function() {
    return _cpp.hasAVX2();
}

exports.hasAltiVec = function() {
    return _cpp.hasAltiVec();
}

exports.hasMMX = function() {
    return _cpp.hasMMX();
}

exports.hasRDTSC = function() {
    return _cpp.hasRDTSC();
}

exports.hasSSE = function() {
    return _cpp.hasSSE();
}

exports.hasSSE2 = function() {
    return _cpp.hasSSE2();
}

exports.hasSSE3 = function() {
    return _cpp.hasSSE3();
}

exports.hasSSE41 = function () {
    return _cpp.hasSSE41();
}

exports.hasSSE42 = function() {
    return _cpp.hasSSE42();
}

exports.default = {};
for (var key in exports) {
    if (key !== 'default') {
        exports.default[key] = exports[key];
    }
}

File: Compute.Core/System.cpp

#include "System.h"
#include "JsExtensions.h"

void battery(v8::FunctionArgs args)
{
    auto secs{ 0 }, pct{ 0 };
    auto status = SDL_GetPowerInfo(&secs, &pct);
    auto object = v8::Object::New(args.GetIsolate());

    AttachNumber(object, "percentage", pct);
    AttachNumber(object, "seconds", secs);
    AttachNumber(object, "status", status);

    args.GetReturnValue().Set(object);
}

void exit(v8::FunctionArgs args)
{
    if (args.Length() == 1 && args[0]->IsNumber())
    {
        exit(args[0].As<v8::Number>()->Value());
    }

    exit(0);
}

void compute::registerSystem(v8::Exports exports)
{
    AttachFunction(exports, "battery", battery);
    AttachFunction(exports, "exit", exit);

    AttachString(exports, "osArchitecture", Poco::Environment::osArchitecture());
    AttachString(exports, "osVersion", Poco::Environment::osVersion());
    AttachString(exports, "osName", Poco::Environment::osName());

    AttachNumber(exports, "cpuCacheLineSize", SDL_GetCPUCacheLineSize());
    AttachNumber(exports, "systemRam", SDL_GetSystemRAM());
    AttachNumber(exports, "cpuCount", SDL_GetCPUCount());

    AttachBoolean(exports, "hasAltiVec", SDL_HasAltiVec());
    AttachBoolean(exports, "hasRDTSC", SDL_HasRDTSC());
    AttachBoolean(exports, "has3DNow", SDL_Has3DNow());
    AttachBoolean(exports, "hasSSE41", SDL_HasSSE41());
    AttachBoolean(exports, "hasSSE42", SDL_HasSSE42());
    AttachBoolean(exports, "hasAVX2", SDL_HasAVX2());
    AttachBoolean(exports, "hasSSE3", SDL_HasSSE3());
    AttachBoolean(exports, "hasSSE2", SDL_HasSSE2());
    AttachBoolean(exports, "hasMMX", SDL_HasMMX());
    AttachBoolean(exports, "hasAVX", SDL_HasAVX());
    AttachBoolean(exports, "hasSSE", SDL_HasSSE());
}